### PR TITLE
Added GitHub Actions workflow to run tests

### LINKED ISSUE #53: Add GitHub Actions workflow to run tests
## Description

As a developer,
I need ensure the tests are passing in the CI pipeline,
so that I can know that my code is not failing before merging.

### Update [Required]
- [ ] Create GitHub actions workflow to run tests

## Definition of Done
- [ ] All of the required items are completed.
- [x] Approval by 1 mentor.

## Estimation
2 hours

### PR DESCRIPTION
**Pull request by codesankalp**
### Description

Added tests.yml to run test written in this project on GitHub workflow.

Changed databases configuration for workflow in `settings.py` to use psql on GitHub workflow.

Fixes #53

### Type of Change:

- Code
- Quality Assurance

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
On GitHub Workflow:
1. Success Build: https://github.com/codesankalp/open-source-programs-backend/actions/runs/560465979
2. Failure Build: https://github.com/codesankalp/open-source-programs-backend/actions/runs/560435054

### Checklist:

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings

### CODE REVIEW

**Comment by codesankalp** [comment]
I didn't find any other method to bypass zulip API key not found error. So I have use mine in workflow.

@isabelcosta Can you please review this pull request?

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        python-version: 3.7
+    - name: Installing dependencies
+      run: |
+        python -m pip install --upgrade pip
+        if [ -f requirements.txt ]; then pip install -r requirements.txt; fi
+    - name: Creating Zulip API key
+      run: |
+        echo "[api]" >> download
+        echo "email=sankalp123427@gmail.com" >> download
+        echo "key=8j7pGWLMuFdb7maOyv0uTTg0nhkQE7Nc" >> download
```
Comment (isabelcosta) [inline .github/workflows/tests.yml:35]: oh no, please remove this. Can you do a little search, we can have environment variables set for GitHub Actions, so you don't need to expose these values. Please squash the commits so this gets taken out

Comment (codesankalp) [reply .github/workflows/tests.yml:35]: OK, @isabelcosta

Comment (isabelcosta) [reply .github/workflows/tests.yml]: here's a little guide about that: https://docs.github.com/en/actions/reference/environment-variables

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    - name: Installing dependencies
+      run: |
+        python -m pip install --upgrade pip
+        if [ -f requirements.txt ]; then pip install -r requirements.txt; fi
+    - name: Creating Zulip API key
+      run: echo ${{ secrets.ZULIP_API_KEY }} >> download
+    - name: Migrating Test Database
+      run: python manage.py migrate
+    - name: Run tests
+      run: python manage.py test --keepdb
```
Comment (devkapilbansal) [inline .github/workflows/tests.yml]: Add a blank line after each step to make it more readable

File: main/settings.py [hunk 3]
```diff
@@ -112,16 +112,30 @@
...
+        "default": {
+            "ENGINE": "django.db.backends.postgresql_psycopg2",
+            "NAME": "osp",
+            "USER": "osp",
+            "PASSWORD": "osp",
+            "HOST": "localhost",
+            "PORT": 5432,
+        }
+    }
+
```
Comment (devkapilbansal) [inline main/settings.py]: Don't make changes here. Instead use these variables in the workflow itself

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    steps:
+    - uses: actions/checkout@v2
+    - name: Set up Python 3.7
+      uses: actions/setup-python@v2
+      with:
+        python-version: 3.7
+    - name: Installing dependencies
+      run: |
+        python -m pip install --upgrade pip
+        if [ -f requirements.txt ]; then pip install -r requirements.txt; fi
```
Comment (devkapilbansal) [inline .github/workflows/tests.yml]: I don't think that you need this condition here. `requirements.txt` file will be there for sure

File: .github/workflows/tests.yml [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+          --health-timeout 5s
+          --health-retries 5
+        ports:
+          - 5432:5432
+    steps:
+    - uses: actions/checkout@v2
+    - name: Set up Python 3.7
+      uses: actions/setup-python@v2
+      with:
+        python-version: 3.7
```
Comment (devkapilbansal) [inline .github/workflows/tests.yml:33]: Can't we use a matrix to test it with different versions

**Review by devkapilbansal — CHANGES_REQUESTED** [review]
Great @codesankalp please look into my request below

File: .github/workflows/tests.yml [hunk 6]
```diff
@@ -0,0 +1,36 @@
+name: Open Source Programs Run Tests
+
+on: [push, pull_request]
+
```
Comment (Amulya-coder) [inline .github/workflows/tests.yml:4]:
```suggestion
branches: [ develop ]
```
Can you add branches also?

File: .github/workflows/tests.yml (hunk 1 above)
Comment (devkapilbansal) [reply .github/workflows/tests.yml]: @isabelcosta better to delete this comment too

File: .github/workflows/tests.yml (hunk 6 above)
Comment (devkapilbansal) [reply .github/workflows/tests.yml:4]: I don't think it is necessary as it should run on all branches of this repository

File: .github/workflows/tests.yml (hunk 1 above)
Comment (codesankalp) [reply .github/workflows/tests.yml]: Yes, I have removed it from commit. It would be better if you delete this comment.

File: .github/workflows/tests.yml (hunk 6 above)
Comment (Amulya-coder) [reply .github/workflows/tests.yml:4]: Ok, so it's by default thanks for clearing :)

File: .github/workflows/tests.yml (hunk 4 above)
Comment (codesankalp) [reply .github/workflows/tests.yml]: I just followed the approach as mentioned in mentor ship backend repo.

File: .github/workflows/tests.yml (hunk 5 above)
Comment (codesankalp) [reply .github/workflows/tests.yml:33]: I tried for stable version.
But any specific reasons for testing for multiple versions?

File: main/settings.py (hunk 3 above)
Comment (codesankalp) [reply main/settings.py]: I followed GitHub action docs to create this container. Is creating env variables for this work?

File: .github/workflows/tests.yml (hunk 2 above)
Comment (codesankalp) [reply .github/workflows/tests.yml]: Sure!

File: .github/workflows/tests.yml (hunk 4 above)
Comment (codesankalp) [reply .github/workflows/tests.yml]: But I also think we don't need checking, so I will remove it. 😊

File: .github/workflows/tests.yml [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    - name: Set up Python 3.7
+      uses: actions/setup-python@v2
+      with:
+        python-version: 3.7
+    - name: Installing dependencies
+      run: |
+        python -m pip install --upgrade pip
+        if [ -f requirements.txt ]; then pip install -r requirements.txt; fi
+    - name: Creating Zulip API key
+      run: echo ${{ secrets.ZULIP_API_KEY }} >> download
```
Comment (devkapilbansal) [inline .github/workflows/tests.yml]: Change it like this

```suggestion
      run: |
          echo "[api]" > download
          echo "email=${{ secrets.ZULIP_MAIL }}" >> download
          echo "key=${{ secrets.ZULIP_API_KEY }}" >> download
          echo "site=https://anitab-org.zulipchat.com" >> download
```

@isabelcosta kindly add `ZULIP_API_KEY` and `ZULIP_MAIL` in repository secrets

**Comment by devkapilbansal** [comment]
@codesankalp you can refer https://github.com/devkapilbansal/open-source-programs-backend/commit/d4db261ba54e9ac35ca3b3157e2808cf81b5c9eb for the changes I asked

File: .github/workflows/tests.yml (hunk 1 above)
Comment (isabelcosta) [reply .github/workflows/tests.yml]: No problem @codesankalp , now you know for next time ;) ! I will delete,

**Comment by codesankalp** [comment]
@isabelcosta , @devkapilbansal Can you please review this again?
I have done the mentioned changes.
Here is the link of build: https://github.com/codesankalp/open-source-programs-backend/runs/1897417544

@isabelcosta Can you create a secret variable named as `ZULIP_API_KEY` and add the API key(add whole content of download file that we get from zulip)  there like this:
image

**Comment by codesankalp** [comment]
> @codesankalp you can refer [devkapilbansal@d4db261](https://github.com/devkapilbansal/open-source-programs-backend/commit/d4db261ba54e9ac35ca3b3157e2808cf81b5c9eb) for the changes I asked

Thanks @devkapilbansal for this. I am missing double quotes yesterday.

**Comment by isabelcosta** [comment]
> @isabelcosta Can you create a secret variable named as ZULIP_API_KEY and add the API key(add whole content of download file that we get from zulip) there like this:

@codesankalp I can do that yes. Can you also add some instructions, either on README or on docs folder markdown file, what we need to have to configure the GitHub Actions, like the env variables needed? Can be a short document.

**Comment by devkapilbansal** [comment]
> @isabelcosta , @devkapilbansal Can you please review this again?
> I have done the mentioned changes.
> Here is the link of build: https://github.com/codesankalp/open-source-programs-backend/runs/1897417544
> 
> @isabelcosta Can you create a secret variable named as `ZULIP_API_KEY` and add the API key(add whole content of download file that we get from zulip) there like this:
> image

@codesankalp add api key and mail id as different variables

**Comment by isabelcosta** [comment]
Is this running code coverage? It would be nice to have an idea of this metric, and see it go up with more tests being added. @codesankalp

**Comment by codesankalp** [comment]
OK, I will add code coverage also @isabelcosta

**Comment by isabelcosta** [comment]
I added API key, do i also need to add mail id? @codesankalp

**Review by isabelcosta — DISMISSED** [review]
I think this is good, I setup the API_KEY secret. Thank you @codesankalp for working on this!

**Comment by devkapilbansal** [comment]
> OK, I will add code coverage also @isabelcosta

Wouldn't it be better to open different issue for coverage

**Comment by isabelcosta** [comment]
@devkapilbansal I guess we could. Do you think this PR is good already? I can merge if you approve

File: .github/workflows/tests.yml [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        options: >-
+          --health-cmd pg_isready
+          --health-interval 10s
+          --health-timeout 5s
+          --health-retries 5
+        ports:
+          - 5432:5432
+    
+    steps:
+    - uses: actions/checkout@v2
```
Comment (devkapilbansal) [inline .github/workflows/tests.yml:29]: There should be a blank line here

File: .github/workflows/tests.yml [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        pip install -r requirements.txt
+    
+    - name: Creating Zulip API key
+      run: echo -e "${{ secrets.ZULIP_API_KEY }}" >> download
+
+    - name: Migrating Test Database
+      run: python manage.py migrate
+
+    - name: Run tests
+      run: python manage.py test --keepdb
```
Comment (devkapilbansal) [inline .github/workflows/tests.yml]: Why you use `keepdb` here?

File: .github/workflows/tests.yml (hunk 5 above)
Comment (devkapilbansal) [reply .github/workflows/tests.yml:33]: Then we may need to add the versions supported in the documentation itself.

File: .github/workflows/tests.yml [hunk 10]
```diff
@@ -0,0 +1,43 @@
+name: Open Source Programs Run Tests
+
+on: [push, pull_request]
+
+jobs:
+  build:
+    name: Run tests
+    runs-on: ubuntu-latest
```
Comment (devkapilbansal) [inline .github/workflows/tests.yml:10]: Add a blank line here too

File: .github/workflows/tests.yml [hunk 11]
```diff
@@ -0,0 +1,43 @@
+name: Open Source Programs Run Tests
+
+on: [push, pull_request]
```
Comment (devkapilbansal) [inline .github/workflows/tests.yml:3]: Is this necessary?? Maybe the actions will work without this line too

**Review by devkapilbansal — CHANGES_REQUESTED** [review]
@codesankalp add GitHub badge in the README too for the build status

File: .github/workflows/tests.yml (hunk 9 above)
Comment (isabelcosta) [reply .github/workflows/tests.yml]: Very good questions 🕵🏾‍♀️

File: .github/workflows/tests.yml (hunk 11 above)
Comment (isabelcosta) [reply .github/workflows/tests.yml:3]: Being explicit is also good. Like by having this explicitly on the workflow, we don't need to think of the default github actions options. but good point

File: .github/workflows/tests.yml (hunk 10 above)
Comment (codesankalp) [reply .github/workflows/tests.yml:10]: > Then we may need to add the versions supported in the documentation itself.

In Readme the mentioned version is python 3.7+ that's why i have used python 3.7 @devkapilbansal 
https://github.com/anitab-org/open-source-programs-backend#tech-stack

File: .github/workflows/tests.yml (hunk 11 above)
Comment (Amulya-coder) [reply .github/workflows/tests.yml:3]: > Is this necessary?? Maybe the actions will work without this line too

@devkapilbansal if we remove this line then how GitHub will know about what event they have to trigger the workflow I think it doesn't let it be a default.

File: .github/workflows/tests.yml (hunk 9 above)
Comment (codesankalp) [reply .github/workflows/tests.yml]: I use --keepdb to preserve the test database in my device. But now I think it's of no use here :sweat_smile: 
I will remove it.

**Comment by codesankalp** [comment]
> I added API key, do i also need to add mail id? @codesankalp

The content of the download (zulip key that you download from settings) file contains the e-mail id.
This must be the content of the secret variable named `ZULIP_API_KEY`:
```
[api]
email=example@example.com
key=********************************************
site=https://anitab-org.zulipchat.com
```
@isabelcosta

File: .github/workflows/tests.yml [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+      with:
+        python-version: 3.7
+    
+    - name: Installing dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install -r requirements.txt
+    
+    - name: Creating Zulip API key
+      run: echo -e "${{ secrets.ZULIP_API_KEY }}" >> download
```
Comment (isabelcosta) [inline .github/workflows/tests.yml:42]: Why is this needed for tests @codesankalp ?

Comment (codesankalp) [reply .github/workflows/tests.yml:42]: The test will not perform without this. Because to perform test there must be no error in code.
And in `utils/zulip_api.py` we are using the `download` file as a config file for Zulip Stats which is a must needed condition before running any command @isabelcosta

**Comment by codesankalp** [comment]
@isabelcosta , @devkapilbansal , @Amulya-coder Can you review this again?

I have done the requested changes. The coverage is also added and it is working good :)
image

Coverage URL: https://codecov.io/gh/codesankalp/open-source-programs-backend/tree/fdeb23da35c5a27f67d341401f3914ce9d6e5073
Build URL: https://github.com/codesankalp/open-source-programs-backend/actions/runs/566612308

I haven't squashed the commit (if needed feel free to comment about squashing, I will squash it :smile: )

**Review by devkapilbansal — DISMISSED** [review]
LGTM :+1:  Great work @codesankalp

**Comment by devkapilbansal** [comment]
@isabelcosta please review it again

File: .github/workflows/tests.yml (hunk 12 above)
Comment (isabelcosta) [reply .github/workflows/tests.yml:42]: Ok, I will approve this. But we have to find a way to not need real API KEY to test, by mocking zulip response when needed, or in last scenario if really needed get a zulip test account. I think we need an issue for this

File: docs/github_actions.md [hunk 13]
```diff
@@ -0,0 +1,15 @@
+# GITHUB ACTIONS
```
Comment (isabelcosta) [inline docs/github_actions.md]:
```suggestion
# GitHub Actions
```

File: docs/github_actions.md [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+
+The purpose of using GitHub actions is to run tests and QA-checks.
+
+## Configuration
+
+There is currently one environment variable needed for configuring the GitHub actions. Details are mentioned below:
+
+1. `ZULIP_API_KEY`: This secret variable is needed to run tests. This is the API key that we get from zulip to access ZULIP API. The content of this secret variable is like this :
```
Comment (isabelcosta) [inline docs/github_actions.md]:
```suggestion
1. `ZULIP_API_KEY`: This secret variable is needed to run tests. This is the API key that we get from zulip to access ZULIP API. To get this key follow t[hese instructions from official documentation](https://zulip.com/api/api-keys#get-your-api-key). The content of this secret variable should look like:
```

**Review by isabelcosta — CHANGES_REQUESTED** [review]
Little change request to improve documentation and then I can approve and merge

**Review by isabelcosta — APPROVED** [review]
Will approve this and hopefully test this as quickly as possible for next PRs to use this.
Thank you @codesankalp for working on this!
